src/dmtest/gendatablocks.py

```diff
@@ -0,0 +1,663 @@
+        value : str
+            the file or path parameter
+
+        Raises
+        ------
+        FileNotFoundError
+
+        """
+        if value is None:
+            return None
+        path = Path(value)
+        if (path.is_file() or path.is_block_device()):
+            return path
+        raise FileNotFoundError(value)
+
+    def verify(self):
+        """Verify the data previously written to a block range.
+
+        Raises
+        ------
+        ValueError
+
+        """
+        if self.path:
+            self.verify_file()
+        else:
+            self.verify_files()
+
+    def verify_file(self):
+        """Verify the data previously written to a block range associated with a file
+        or device.
+
+        Raises
+        ------
+        ValueError
+
+        """
+        if self.path is None:
+            raise ValueError("the file/device path is invalid")
+
+        flags = os.O_RDONLY
+        with os.fdopen(os.open(self.path, flags), "rb") as fd:
+            self._seek(fd)
+            for n in range(0, self.block_count):
+                data = fd.read(self.block_size)
+                self.verify_streams(n, data)
+
+    def verify_files(self):
+        """Verify the data previously written to a block range associated with a
+        directory.
+
+        Raises
+        ------
+        ValueError
+
+        """
+        if self.dir_path is None:
+            raise ValueError("the directory path is invalid")
+
+        os.chdir(self.dir_path)
+        self.iterate_over_files(self.streams[0], False)
+
+    def verify_streams(self, block_number: int, actual: bytes):
+        """Verify all streams related to a specific block in a block range.
+
+        Parameters
+        ----------
+        block_number : int
+            block number to verify all streams against
+        actual : bytes
+            the bytes to compare against
+
+        Raises
+        ------
+        CompareError
+
+        """
+        for stream in self.streams:
+            if stream.claim(actual):
+                expected = stream.generate(block_number, self.block_size)
+                if expected == actual:
+                    stream.counter += 1
+                    return
+                for i in range(0, self.block_size):
+                    if actual[i] != expected[i]:
+                        raise CompareError("block compare failure",
+                                           block_number, actual, expected, i)
+        raise CompareError("block not recognized.", block_number, actual)
+
+    def write(self,
+              tag: str,
+              dedupe: float = 0.0,
+              compress: float = 0.0,
+              direct: bool = False,
+              sync: bool = False,
+              fsync: bool = False):
+        """Write to a block range
+
+        Parameters
+        ----------
+        tag : str
+            tag the data for future reference
+        dedupe : float
+            how much dedupe to write
```

Comment:
how much duplication

Reply:
done

Reply:
again, not DEduplication ... the data generated has duplicates